web/packages/shared/hooks/useRefClickOutside/useRefClickOutside.ts

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { MutableRefObject, useEffect, useRef, useCallback } from 'react';
+
+/**
+ * useRefClickOutside adds a `mousedown` event listener (and cleanup)
+ * that upon `mousedown` will:
+ *   - set the field `open` to false if the event happened outside the ref
+ *   - does nothing if the event happened inside the ref
+ *
+ * Returns a `ref` object to be used for the element that we want `mousedown`
+ * events to be ignored.
+ */
+export function useRefClickOutside<
+  T extends { contains(eventTarget: HTMLElement): boolean }
+>(options: { open: boolean; setOpen(b: boolean): void }): MutableRefObject<T> {
+  const ref = useRef<T>();
+
+  const handleClickOutside = useCallback((event: MouseEvent) => {
+    if (ref.current && !ref.current.contains(event.target as HTMLElement)) {
+      options.setOpen(false);
+    }
+  }, []);
```

Comment:
I suppose `setOpen` should be added to the deps here, no? Otherwise even if it changes, `handleClickOutside` will keep calling the old function.

<details>
<summary>Patch</summary>

```diff
diff --git a/web/packages/shared/hooks/useRefClickOutside/useRefClickOutside.ts b/web/packages/shared/hooks/useRefClickOutside/useRefClickOutside.ts
index 4ca33d5ea7..3dbf9cacec 100644
--- a/web/packages/shared/hooks/useRefClickOutside/useRefClickOutside.ts
+++ b/web/packages/shared/hooks/useRefClickOutside/useRefClickOutside.ts
@@ -29,22 +29,26 @@ export function useRefClickOutside<
   T extends { contains(eventTarget: HTMLElement): boolean }
 >(options: { open: boolean; setOpen(b: boolean): void }): MutableRefObject<T> {
   const ref = useRef<T>();
+  const { setOpen, open } = options;
 
-  const handleClickOutside = useCallback((event: MouseEvent) => {
-    if (ref.current && !ref.current.contains(event.target as HTMLElement)) {
-      options.setOpen(false);
-    }
-  }, []);
+  const handleClickOutside = useCallback(
+    (event: MouseEvent) => {
+      if (ref.current && !ref.current.contains(event.target as HTMLElement)) {
+        setOpen(false);
+      }
+    },
+    [setOpen]
+  );
 
   useEffect(() => {
-    if (options.open) {
+    if (open) {
       document.addEventListener('mousedown', handleClickOutside);
 
       return () => {
         document.removeEventListener('mousedown', handleClickOutside);
       };
     }
-  }, [ref, options.open, handleClickOutside]);
+  }, [open, handleClickOutside]);
 
   return ref;
 }
```
</details>